src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/CrossChainTransferStore.cs

```diff
@@ -177,155 +204,234 @@ private HashHeightPair TipToChase()
+                    this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
+                        this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            }
 
-                using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            var tracker = new StatusChangeTracker();
+            foreach (CrossChainTransfer partialTransfer in crossChainTransfers)
+            {
+                // Verify that the transaction input UTXO's have been reserved by the wallet.
+                if (!SanityCheck(partialTransfer.PartialTransaction, wallet, partialTransfer.Status == CrossChainTransferStatus.FullySigned))
                 {
-                    dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
+                    tracker.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
+                }
+            }
 
-                    CrossChainTransfer[] partialTransfers = this.Get(dbreezeTransaction,
-                        this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
-                            this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            if (tracker.Count == 0)
+                return;
+
+            using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            {
+                dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
 
-                    foreach (CrossChainTransfer partialTransfer in partialTransfers)
+                try
+                {
+                    foreach (KeyValuePair<ICrossChainTransfer, CrossChainTransferStatus?> kv in tracker)
                     {
-                        if (!SanityCheck(partialTransfer.PartialTransaction, wallet))
-                        {
-                            this.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
-                            await this.PutTransferAsync(dbreezeTransaction, partialTransfer);
-                        }
+                        this.PutTransfer(dbreezeTransaction, kv.Key);
                     }
 
                     dbreezeTransaction.Commit();
+
+                    this.UpdateLookups(tracker);
+                }
+                catch (Exception err)
+                {
+                    this.logger.LogError("Error during database update: {0}", err.Message);
+                    this.logger.LogTrace("(-):[SANITY_ERROR]");
+
+                    // Restore expected store state in case the calling code retries / continues using the store.
+                    dbreezeTransaction.Rollback();
+                    throw err;
                 }
             }
+        }
 
-            this.logger.LogTrace("(-)");
+        public static Transaction BuildDeterministicTransaction(uint256 depositId, Recipient.Recipient recipient, Network network,
```

Comment:
`network` not used.

Reply:
The partials argument has since been removed. It is deterministic due to the UTXO ordering taking place in the `federationWalletTransactionHandler.BuildTransaction` method. Also we are highlighting the fact that the transaction must be built in a specific way - i.e. a single recipient plus specific arguments in `multiSigContext`.

Reply:
Ok, understand.  I would say `federationWalletTransactionHandler.BuildTransaction` becomes `federationWalletTransactionHandler.BuildDetermisticTransaction`.  And `CrossChainTransferStore.BuildDeterministicTransaction` becomes `CrossChainTransferStore.BuildMultiSigTransaction` or `CrossChainTransferStore.BuildFederatedTransaction`.  But probably fine as it is.

Reply:
I've added `OrderCoinsDeterministic` to the `TransactionBuildContext`. Since `federationWalletTransactionHandler.BuildTransaction` can fulfill either the deterministic or non-deterministic role I am leaving the name unchanged.